wallet/chainntfns.go

```diff
@@ -345,30 +356,65 @@ func (w *Wallet) addRelevantTx(dbtx walletdb.ReadWriteTx, rec *wtxmgr.TxRecord,
-	// We'll start by acquiring our chain backend client as we'll be
-	// querying it for blocks.
-	chainClient, err := w.requireChainClient()
-	if err != nil {
-		return nil, err
-	}
+// chainConn is an interface that abstracts the chain connection logic required
+// to perform a wallet's birthday block sanity check.
+type chainConn interface {
+	// GetBestBlock returns the hash and height of the best block known to
+	// the backend.
+	GetBestBlock() (*chainhash.Hash, int32, error)
 
-	// We'll then fetch our wallet's birthday timestamp and block.
+	// GetBlockHash returns the hash of the block with the given height.
+	GetBlockHash(int64) (*chainhash.Hash, error)
+
+	// GetBlockHeader returns the header for the block with the given hash.
+	GetBlockHeader(*chainhash.Hash) (*wire.BlockHeader, error)
+}
+
+// syncStore is an interface that abstracts the wallet's sync-related
+// information required to perform a birthday block sanity check.
+type syncStore interface {
+	// Birthday returns the birthday timestamp of the wallet.
+	Birthday() time.Time
+
+	// BirthdayBlock returns the birthday block of the wallet.
+	BirthdayBlock(walletdb.ReadBucket) (waddrmgr.BlockStamp, bool, error)
+
+	// SetBirthdayBlock updates the birthday block of the wallet to the
+	// given block. The boolean can be used to signal whether this block
+	// should be sanity checked the next time the wallet starts.
+	SetBirthdayBlock(walletdb.ReadWriteBucket, waddrmgr.BlockStamp, bool) error
+
+	// SetSyncedTo marks the address manager to be in sync with the
+	// recently-seen block described by the blockstamp. When the provided
+	// blockstamp is nil, the oldest blockstamp of the block the manager was
+	// created at and of all imported addresses will be used.  This
+	// effectively allows the manager to be marked as unsynced back to the
+	// oldest known point any of the addresses have appeared in the block
+	// chain.
+	SetSyncedTo(walletdb.ReadWriteBucket, *waddrmgr.BlockStamp) error
+}
+
+// birthdaySanityCheck is a helper function that ensures a birthday block
+// correctly reflects the birthday timestamp within a reasonable timestamp
+// delta. It's intended to be run after the wallet establishes its connection
+// with the backend, but before it begins syncing. This is done as the second
+// part to the wallet's address manager migration where we populate the birthday
+// block to ensure we do not miss any relevant events throughout rescans.
+func birthdaySanityCheck(db walletdb.DB, chainConn chainConn,
+	syncStore syncStore) (*waddrmgr.BlockStamp, error) {
+
+	// We'll start by fetching our wallet's birthday timestamp and block.
 	var (
-		birthdayTimestamp     = w.Manager.Birthday()
+		birthdayTimestamp     = syncStore.Birthday()
 		birthdayBlock         waddrmgr.BlockStamp
 		birthdayBlockVerified bool
 	)
-	err = walletdb.View(w.db, func(tx walletdb.ReadTx) error {
+
+	err := walletdb.View(db, func(tx walletdb.ReadTx) error {
```

Comment:
Why not throw this into an interface above as well? No need to create an entire mock wallet db in this case to test the function.